### ISSUE TITLE
Compiler API: no Symbol for Node

### ISSUE THREAD

**Issue by aiboost** [microsoft/TypeScript#13165]
**TypeScript Version:**  2.1.1

**Code**

I'm trying to implement function detecting is identifier internal (can be renamed) or external but sometimes TS Compiler API returns me no symbols for identifiers:
```ts
private isIdentExternal(node: ts.Node): boolean {
    // node here is only ts.Node with kind=65 (Identifier), and already checked not to be reserved word
    let symbol = this.program.getTypeChecker().getSymbolAtLocation(node);
    if (!symbol) {
        // why?
    }
    for (const decl of symbol.declarations)
      if (decl.getSourceFile().fileName.endsWith(".d.ts"))
        return true;
    return false;
  }
```

There are two known cases when I get this problem.

Case 1: no symbol for internal abstract function:
```ts
abstract class Cls {
    private abstract func();// no symbol is here
    // Tough I can get symbols from cls.func() code in other place, the problem is for declaration only
}
```

Case 2: no symbol for external properties:
```ts
let el: HTMLElement = document.createElement("div");
el.className = "x"; // no symbol for "className", tough I have symbol for "createElement" (problem is only for properties)
```

**Expected behavior:**

Every identifier (function/variable/method/property/class definition) has symbol.

**Actual behavior:**

Sometimes symbol is missing (TypeChecker does not return it).

**Comment by dmiko** [comment]
Any hint? 
Please tell me where I am wrong and is this actually a bug?
I've found some more cases with no Node->Symbol conversion, and do not understand how can it be possible. I need it to write custom obfuscator for my app.

**Comment by voldmar** [comment]
Could you also give SourceFile to your function as argument and use it?

**Comment by dmiko** [comment]
@voldmar , yes, I can pass SourceFile there but this will be source file of place where the call exists, not where the definition is made.

For example in the code ````el.className```` SourceFile will be always my source .ts, but the symbol ````className```` is defined in "es5.d.ts"

**Comment by voldmar** [comment]
Hm, I just have used `ts.ModuleKind.ES2015` for module type and got proper `el.className` resolution

**Comment by dmiko** [comment]
@voldmar , I have these options for compiler:

```ts
const options: ts.CompilerOptions = {
  allowNonTsExtensions: true,
  module: ts.ModuleKind.AMD,
  target: ts.ScriptTarget.ES5,
};
```

ES5 is passed to compiler fine (methods have symbols).
Looks like maybe I need to play with compiler version... In any case thank you for advice.

**Comment by pronebird** [comment]
@RyanCavanaugh any progress on this? I have a similar issue in a custom transformer that replaces all calls to `process.env.NODE_ENV` with a string literal. Basically `getSymbolAtLocation` works very well but there are edge cases where it's useless, for instance it returns `undefined` for `process['env']['NODE_ENV']`, but works for `process.env.NODE_ENV`...

```ts
const ts = require('typescript');

module.exports = function (program) {
  const typeChecker = program.getTypeChecker();

  return function (context) {
    const visit = (node) => {
      const isPropertyExpr = ts.isPropertyAccessExpression(node) && node.name.text === 'NODE_ENV';
      const isElementExpr = ts.isElementAccessExpression(node) && node.argumentExpression.text === 'NODE_ENV';

      if (isPropertyExpr || isElementExpr) {
        const symbol = typeChecker.getSymbolAtLocation(node.expression);

        if (!symbol) {
          throw new Error('No symbol for ' + node.getFullText());
        }

        if (symbol && symbol.getName() === 'env' && symbol.parent && symbol.parent.getName() === 'Process') {
          return ts.createLiteral('production');
        }
      }

      return ts.visitEachChild(node, (child) => visit(child), context);
    };
    return (node) => ts.visitNode(node, visit);
  };
};

```

**Comment by foxel** [comment]
Any update on this issue? I'm facing the same while implementing a custom rule for tslint.

**Comment by lolleko** [comment]
Also facing a similar issue. Would appreciate some update.

**Comment by Perryvw** [comment]
Also ran into this today. Surprisingly, while comparing two functions in the same source file, one of them has a symbol and the other does not.

[function startParty()](https://github.com/ModDota/dota-tutorial/blob/master/game/scripts/vscripts/Sections/Chapter6/SectionClosing.ts#L430) does not have a symbol in the compiler API.

A few lines down, [function stopParty()](https://github.com/ModDota/dota-tutorial/blob/master/game/scripts/vscripts/Sections/Chapter6/SectionClosing.ts#L460) has a symbol as expected.

These functions are both pretty simple and very similar, making this bug even weirder.  We have tried but were not able to create a smaller reproduction.

**Comment by DanielSWolf** [comment]
I'm experiencing the same problem trying to get the symbol for a variable declaration. I tried `getSymbolAtLocation()` with the `VariableDeclaration`, the `VariableStatement`, and, out of sheer desperation, the `VariableDeclarationList`. In each case, I only got `undefined`.

I'm not sure if that's any help, but the AST explorer shows the same behavior. I just typed `let foo: number;` and expanded the AST. Under `VariableDeclaration`, it shows `symbol: undefined` (see screenshot). Is there some option I failed to set?

![image](https://user-images.githubusercontent.com/5963162/134895304-55c9dfbc-e178-4954-8fcf-dba491a25de7.png)

**Comment by RyanCavanaugh** [comment]
This concerns the pre-TypeScript-7 Compiler API. That API, including `TypeChecker.getSymbolAtLocation`, has been superseded by the TypeScript 7 API and is no longer being developed. Closing this request accordingly.
